Initial merge bot implementation.

eduard-bagdasaryan opened the pull request:
PR merging with auto-branch implementation.

This branch fully reworks the initial bot version. Here are some of the ideas it follows to:

* Do not cache any information received from Github, except data used for processing a single PR.

* Do not use received events data, since it is incomplete and requires caching for its processing.

* Use events only as triggers to start polling all available open PRs.

* Use a special 'auto_branch' reference in order to mark being-in-merge commit.
   This commit is created as a merge product of the PR and its base branch (master).

* Merge only one PR at a time, all other PRs should wait until it is successfully merged or failed to 
 merge(e.g., auto_branch checks failed).

* Do all repository manipulations with Github API in order to avoid tricky working with the cloned local copy.

* Automate as much as possible. For example, the bot should itself decide whether to re-try merging previously failed PRs.

eduard-bagdasaryan committed:
Fixed eslint errors except 'defined but never used'.

The 'no-unused-vars' rule currently set to 'error'.  There are some
callbacks passed to another libraries, having such variables. I don't
think we should change passed callbacks signature, because these
variables (e.g., error messages) are unused for now, but may be used in
future.

eduard-bagdasaryan committed:
Polishing: substituted with arrow function expressions.

eduard-bagdasaryan committed:
Reworked eggtimer bot to conform to requirements.

AFAICT, the requirements are:

1. Do not cache any information received from Github, except data used
   for processing a single PR.

2. Do not use received events data, since it is incomplete and requires
   caching for its processing.

3. Use events only as triggers to start polling all available open PRs.

4. Configure Github repo/owner in order to do PRs polling on bot
   startup.

This revision does not change 'merge' functionality yet (including
auto_branch operation).

eduard-bagdasaryan committed:
Initial 'PR merging via auto_branch' implementation.

Not fully tested yet.

eduard-bagdasaryan committed:
Fixed some bugs after testing.

eduard-bagdasaryan committed:
Re-worked merging algorithm, using Github 'commit' API.

The previous implementation, which was based on Github 'merge' API,
has a flaw: after merge into auto_branch is done, there is no way
to undo it. It makes no problem if auto_branch tests are successful,
but if they are not, we end up with closed PR and inability to
finish our merge into master.

There is no ideal solution for this problem, so our trade of is
the following: do not use 'merge' button API and do manual merge
instead. The disadvantage is after merging the PR gets only
'closed' status (instead of 'merged' and 'closed'), confusing the
reader that the PR became abandoned for some reason. To eliminate
this problem we decided to mark the PR with a label(not done yet).

This solution makes use of Github 'merge' revision, stored as
a reference under 'refs/pull/:pr/merge' reference path, which
is created by Github for each PR. With Github 'commit' API,
we create a new commit, pointing to this 'merge' revision and reset
auto_branch into it. If checks are successful, we merge master
with FF mode into 'auto', closing the PR.

eduard-bagdasaryan committed:
Mark closed PRs with "S-merged" label.

eduard-bagdasaryan committed:
Mark merging PR which auto checks were failed.

If PR's auto branch checks failed during merge, the PR is has still
'mergeable' status and is scheduled again and again.  I mark such PRs
with a label 'S-merge'failed' in this case.  When the underlying
problems are fixed, this label can be removed manually, thus activating
PR merging again.

Also: unified PR numbering in debug messages.

eduard-bagdasaryan committed:
Do not use Github 'merge' revision, create it by hand.

Using 'merge' revision, supplied by Github does not work if master
changes: we don't want to bother user pressing "Update branch" on the
PR. So I have to create that 'merge' revision separately, using master's
HEAD for that.

eduard-bagdasaryan committed:
Revert "Do not use Github 'merge' revision, create it by hand."

This reverts commit c2a42388bebd45e8867548c495c2516a3a28dc4d.
That change was based on the incorrect assumption that PR 'merge'
revision is not updated if master changes.

eduard-bagdasaryan committed:
Reworked non-auto_branch code with promises.

With this refactoring, the code looks simpler and more reliable.

eduard-bagdasaryan committed:
Restore merging context on startup.

Now, when bot starts, it checks whether there is a still merging
auto_branch revision. If there is one, it completes merging process.
This is needed primary for a situation when the bot was terminated while
it was waiting for auto_branch checks completing (for a PR had been
merged into auto_branch).

eduard-bagdasaryan committed:
Restore merging context on startup (pt. 2).

Added functionality for commenting being-in-merge commit and dropping
this comment after merge into master is completed.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Properly handle previous unsuccessful merge attempts.

Before this change, such PRs (which are marked with "S-merge-failed"
label) were ignored, until somebody drops this label manually.
Now, a PR with "S-merge-failed" is considered as a merge candidate if
its 'merge' revision has been changed since last unsuccessful merge
attempt (in hope the underlying problem has been fixed by these
changes).  Github automatically updates 'merge' revision if
changes either PR branch or its base branch(usually master).

eduard-bagdasaryan committed:
Fixed several problems after testing.

* PR 'merge' commit is not updated automatically when
  master changes. To make this happen, we should get the PR via API.

* Distinguish 'ref' parameter in
  createReference/updateReference/deleteReference API calls.

* Do not reset 'currentContext' before its 'signaled' flag is processed.

* Properly manage with 'merging-PR' tag: update if necessary and delete
  after merge completion.

* Few other bugs.

eduard-bagdasaryan committed:
Updated config template with actual parameters.

eduard-bagdasaryan committed:
More tests performed and fixed found problems.

* When getting PR information about 'mergeable' flag,
  this information may be still unavailable, meaning that Github is
  currently calculating this 'mergeability'. In this case, we should
  repeat request, until Github returns non-null 'mergeable'.

* Remove "S-merge-failed" label after problems are fixed and the PR
  is merged.

eduard-bagdasaryan committed:
Refactoring: use 'merging tags' instead of 'merging commit comments'.

Both mechanisms were excessively used for marking such 'being-in-merge'
commits. This marking is needed to be able to restore merging context if
something goes wrong during merge, for example, bot terminates or PR
checks fail. This change eliminates commit commenting, because tagging
is enough.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+        }
+        console.log("Will process", res.data.length, "open PRs");
+        PRList = res.data;
+        processNextPR();
```
I find the getPRList-processNextPR-getPRList "recursion" or "loop" confusing. It is difficult for me to convince myself that this loop is correct and necessary. The presence of the processNextPR() parameter that significantly alters the nature of that function does not help either. Please try to refactor the code to replace processEvent() (which does not actually process any events) with something close to the following run() function:

```
let Running = false; // run() loop is currently running
let Rerun = false; // abort the current PR advancement attempt

// process as many pull requests as possible
async function run()
{
    if (Running) {
        console.log("planning to redo the current run() iteration to account for concurrent GitHub changes");
        Rerun = true; // may already be true
        return;
    }

    Running = true;
    while (await runStep() || Rerun) {
        console.log("finished one run() iteration");
        Rerun = false; // reset; there is nothing to rerun yet
    }
    Running = false;
}

// find and advance (towards merging) the next pull request
// returns true if and only if something was changed
async function runStep()
{
    const pullRequests = await getPullRequests();
    const toAdvance = await findPullRequestToAdvance(pullRequests);
    if (toAdvance)
        return await advancePullRequest(toAdvance);
    return false;
}
```

Please note that run() above is the only place where runStep() is called and  runStep() is the only place where getPullRequests() and advancePullRequest() are called. When advancePullRequest() promise is resolved (i.e., the identified pull request is advanced to testing or merged), EggTimer starts from scratch.

getPullRequests(), advancePullRequest(), and/or some lower-level blocking functions should abort if they detect a true Rerun global. However, I do not expect many places to check for Rerun. Ultimately, only one such place is necessary -- where we issue the final merge command to update master. The rest are optimizations.

eduard-bagdasaryan replied:
```
I find the getPRList-processNextPR-getPRList "recursion" or "loop" confusing. It is difficult for me to convince myself that this loop is correct and necessary.
```
If my understanding is correct, there is one primary difference between my approach and yours (apart from your 'recursion' constructions dislikes). I request all available PRs at once and process them sequentially, whereas you suggest requesting all available PRs after each PR processing step.

I see two reasons why my approach is better:

* it is more optimized: if there are no changes on Github(i.e., PR list is the same), we keep processing the cached list. If there are, we will know about that due to an event and request the full list again.

* Imagine a situation if an all_green_PR failed while auto_branch testing. With your approach, we will keep requesting this PR again and again, until the problems are fixed. With mine, we will wait for an event(which is probably a sign that problems have been fixed).

rousskov replied:
> With your approach, we will keep requesting this PR again and again

This is not my approach.

> [My approach is] optimized: If there are no changes on Github, we keep processing the cached list.

Agreed. However, it would be easy to add that optimization to my approach as well if needed. It is a minor detail, not a key difference.

> apart from your 'recursion' constructions dislikes

The lack of "recursion" is one of the primary differences. The lack of global state (currentContext) is another. There is also the fact that, in my scheme, the candidate PR is always chosen by a central method, based on all available information.

One sketch aspect that I am not sure about is the concept of advancing the chosen PR to the extent possible and then re-evaluating the choice (and possibly advancing another PR). The alternative would be to get rid of the "advancing" concept and simply merge the chosen PR:

```
// find and merge the next pull request
// returns true if and only if something was changed
async function runStep()
{
    const pullRequests = await getPullRequests();
    const toMerge = await findPullRequestToMerge(pullRequests);
    return toMerge && await mergePullRequest(toMerge);
}
```

In this simpler scheme, mergePullRequest() blocks until the request is merged (including waiting for auto tests!) or rejected for some reason. This scheme is better (because it is simpler) if there is (and will be) nothing a bot can do to pull request B while advancing pull request A. I do not think that precondition will be true for long because I expect the bot to be able to, say, auto-format PR branches (while merging another PR) or merge multiple PRs (with multiple target branches) at the same time.

However, I think we can refactor the sketched high level code a little to support concurrent PR action when needed virtually _without_ changing the underlying low-level code. The key to that refactoring would be in moving global `Running` concurrency protection from run() to functions inside runStep() (e.g., the sketched findPullRequestToMerge() or the future findPullRequestToFormat()).

Thus, I recommend using this simpler mergePullRequest() model as opposed to the more complex advancePullRequest() one!

eduard-bagdasaryan committed:
Refactored and improved logging.

* Unifying messages, generated after promisificated methods completion
  (i.e., 'resolve' or 'reject' detailed information).

* Several new helper methods and objects used to collect and output
  exception messages.

* Improved messages wording.

eduard-bagdasaryan committed:
Refactoring: grouped related methods, no functionality changes.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+        .then((result) => {
+             assert(result === true);
+        })
+        .catch((err) => {
```
It feels like there are too many error handling (.catch) places in the code, which probably also means that there are not enough (i.e., some cases are missed). AFAICT, there should be no need to catch all exceptions close to their origin. Is it possible to unify handling of all exceptions at the top (i.e., processEvent() or similar) level?

eduard-bagdasaryan replied:
```
Is it possible to unify handling of all exceptions at the top (i.e., processEvent() or similar) level?
```

I was thinking about that: in current implementation this is possible if convert 'startMerging()' and 'checkMergePreconditions()' into promises and make a high-level try-catch. I doubt, however, that this would be a valuable optimization. finishMerging() is handled separately: it is activated if a auto_status has come.

rousskov replied:
> finishMerging() is handled separately: it is activated if a auto_status has come

As we discussed, this will not be true in the next revision. Perhaps that will help you centralize error handling. Still, there are different _kinds_ of errors (with different effect on GitHub and bot states) so error handling will not be trivial even if it is centralized.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+            return;
+        })
+        .catch((err) => {
+            console.log("No PR is in the merging state: ", err);
```
Lack of mergeable PRs is not really an error or exceptional condition. If possible, please refactor the code to treat this condition normally. My runStep() sketch in another review comment may be useful for that -- the `pullRequests` array there will be empty and/or `toAdvance` will be nil if there are no pull requests to advance.

Please check other .catch statements for similar problems. They should all handle exceptional situations.

eduard-bagdasaryan replied:
Though this is an exception, it is not an error, because I used 'console.log()' instead of 'console.error()'. Using exceptions helps to stop later processing and exit immediately, I don't understand what is bad in using them.

rousskov replied:
Using exceptions for normal/desirable/anticipated/regular conditions conflicts with the exception intended purpose/design. Such misapplication of standard development concepts may confuse developers, make code more complex than necessary, slow down the program, and lead to bugs. We should use exceptions for their intended purpose unless there is a very good (and clear or documented) reason to violate this natural rule of thumb.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+WebhookHandler.on('status', (ev) => {
+    const e = ev.payload;
+    console.log("status event:", e.id, e.sha, e.context, e.state);
+    if (merging(e.sha)) {
```
If this check is an optimization, please remove this check (for now). We will optimize later if needed.

eduard-bagdasaryan replied:
No, that is not an optimization. This handles status events, coming from auto_branch tests.

rousskov replied:
If this is not an optimization, then this is a bug: The bot should not rely on receiving any specific notification (because, for example, the bot may not be running when GitHub sends the notification).

eduard-bagdasaryan replied:
I don't see where exactly this bug is: we do not rely _only_ on this information. The situation when the bot was off while such events came is handled separately.

rousskov replied:
> I don't see where exactly this bug is: we do not rely _only_ on this information.

The question is _not_ whether we rely on some other information as well. The question is whether we rely on (i.e., require) _this specific event_ to function correctly.

* If we rely on it, then it is a bug because it violates the "do not rely on receiving any specific notification" invariant.
* If we do not rely on it, then it is just an optimization (which may be safely removed to simplify the code; for now).

Are you implying that "do not rely on receiving any specific notification" is a wrong invariant to uphold? I do not think you do. Given the fact that GitHub implementation itself and bot access to the internet are not 100% reliable, I do not see why that invariant is wrong...

You have attacked an example where that invariant is important (a bot server crash) because that example is irrelevant to this use case, but that was just the simplest example to illustrate the invariant. In our deliberations, we should focus on the invariant itself.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
-    res.end('no such location')
-  });
-}).listen(CONFIG.port);
+const Config = JSON.parse(fs.readFileSync('config.js'));
```
Please either create a Config class (with methods named after parameters) to make sure that all typos when accessing configuration values automatically result in errors/exceptions. 

Using JSON for the raw configuration file syntax (and for loading raw values in the class constructor) is fine, of course.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+}
+
+function markedAsFailed(labels) {
+    if (labels === undefined)
```
Undefined labels is either a markedAsFailed() caller bug or an implementation bug somewhere else -- the code should not be asking whether an undefined container contains a particular element! If anything, we could assert() that labels are defined here (but that is not necessary because we are dereferencing them). A labels container can be empty, of course.

Please check for similar problems in other label-checking functions.

eduard-bagdasaryan replied:
Agreed, this was a stale code.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+    let statusParams = commonParams();
+    assert(prContext.autoSha);
+    statusParams.ref = prContext.autoSha;
+    let errorLabel = MergeFailedLabel;
```
If `MergeFailedLabel` should be set at all (discussed elsewhere), then it should be set when we fail to either declare the PR not ready for merging or merge it. The finishMerging() function covers only some stages of that long process of advancing a PR towards a merger. It feels wrong that we are setting `MergeFailedLabel` here as opposed to some higher level code that would cover all the applicable stages.

I also suspect that we need to be more careful with promise rejections. Among all the non-mergeable cases, there should be a clear distinction between "not yet ready for merging" cases and "failed to merge" cases. Both result in the higher level code looking at the next PR candidate, but only the latter may label the PR and/or add a GitHub comment to bring human attention to the matter. If we want a higher level .catch() statement to be able to distinguish the two cases, then perhaps we should reject not-yet-ready-PRs specially while always throwing Error objects on errors?

eduard-bagdasaryan replied:
```
If MergeFailedLabel should be set at all (discussed elsewhere), then it should be set when we fail to either declare the PR not ready for merging or merge it.
```

In my understanding, we should set this label in PR if and only if the last stage (finishMerging()) is OK, actually meaning that auto tests have been failed.
However, I probably I am misunderstanding this question. Note, that in bfc2af0 I have changed that code, adding a simple flag to distinguish two situations in finishMerging():
* Reject because not all auto_checks have been completed (without PR labeling)
* Reject on an error or failed auto_checks (with PR labeling).

rousskov replied:
Let's revisit this after the next rewrite. Just keep in mind that 

* we want humans to be able to see important PR states at any given time _but_ we do not want to bombard humans with temporary, unimportant, and/or duplicated information. 
* the bot should not get into a livelock situation where the bot repeats the same (failing) action over and over again.

From those principles point of view, a `MergeFailedLabel` that only reflects the "last stage is OK, [but] auto tests have failed" state is probably not needed because the state of auto tests is already visible on human-oriented GitHub pages and is available to the bot via GitHub API -- no special label is needed to reflect all that.

My "fail to either declare the PR not ready for merging or merge it" conditions are more general and do not contradict the above principles. In the more general context, "merging" should be interpreted as "advancing" (i.e., making progress towards merging or actually merging).

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+const Github = new nodeGithub({ version: "3.0.0" });
+const GithubAuthentication = { type: 'token', username: Config.github_username, token: Config.github_token };
+
+const MergeFailedLabel = "S-merge-failed";
```
I do not understand why we need this label. It does not seem to be used to skip failed PRs (i.e., it is logged but otherwise ignored if found). If the bot does not need it, we should probably remove it.

eduard-bagdasaryan replied:
This label is needed for information purposes only (as we decided during latest discussion). I decided to leave it since it looks useful.

The JIRA issue requirement:

``` 
Eventually, the bot should probably update the status of the PR as well (to reflect auto tests)
```

rousskov replied:
A label marking stuck/unmergeable PRs would be useful indeed. In fact, it may be _required_ for the correct bot operation. This is not such a label AFAICT because it marks transient, not permanent failures (i.e., it is ignored/reset by the bot).

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+       .then((results) => {
+           let commitObj = results[0];
+           let pr = results[1];
+           assert(pr.number = prContext.pr.number);
```
Please eliminate suspicion that you meant `===`.

eduard-bagdasaryan replied:
Fixed.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+
+function initContext() {
+    assert(currentContext === null);
+    currentContext = {};
```
The difference between the current context and the context(s?) created by createPRContext() is not clear to me.

Please convert context into a proper class and clarify (or eliminate) the difference between the two sets. If "context" is meant to represent a pull request being processed, then perhaps the class should be called PullRequest. If `context.pr` is pull request returned by the GitHub API/libraries, then either the two should be merged or the latter can be stored as a "raw" data field of the former. The comment about Config field name typos detection are relevant here as well.

eduard-bagdasaryan replied:
I have already removed 'other contexts' in bfc2af0. The only context left is 'currentContext', i.e., data related to a current being-in-process pull request.

rousskov replied:
Please convert currentContext into a proper PullRequest (or similar) class. If context.pr is pull request returned by the GitHub API/libraries, then either the two should be merged or the latter can be stored as a "raw" data field of the former. The comment about Config field name typos detection are relevant here as well.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+                if (!(st.context in checks)) {
+                    if (st.state !== 'pending') {
+                        console.log("adding", st.context, st.state);
+                        checks[st.context] = (st.state === 'success');
```
I think we should store the Status object (or similar) instead of the result of a "the test was successful" check. Yes, doing so will not allow you to use the checkValues() hack, but I suspect we will need more status information soon and, more importantly, I am uncomfortable with that hack and associated type erasure.

Alternatively, getStatuses() could be converted to produce the final/analyzed result instead of raw booleans, so that getStatuses() callers do not need to iterate over statuses at all (to find out whether all the required tests have been successful).

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+           params.ref = "pull/" + prContext.pr.number + "/merge";
+           return getReference(params);
+       })
+       .then((sha) => {
```
I suspect this `sha` is not the same SHA as the `sha` in the earlier `.then( (sha))` clause. Please give both SHAs more specific/meaningful names.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+           params.sha = sha;
+           return getCommit(params);
+       })
+       .then((obj) => {
```
I suspect this `obj` is a commit object but not the same commit object as the `commitObj` in the earlier `.then( (results))` clause. Please give this commit object a more specific/meaningful name than `obj`.

eduard-bagdasaryan replied:
I was thinking about these results naming, but renaming obj->commitObj does not give any valuable information, because:
1. Evidently, in such 'then' we are getting the result of the last promisificated method (getCommit) in our case, i.e., this 'obj' is its result.
2. Since bfc2af0 almost all promisificated methods return a 'hash object'; we should look at it anyway in order to use it.

rousskov replied:
> renaming obj->commitObj does not give any valuable information

I did not suggest that renaming and certainly do not insist on it. A part of my original concern was (and still is) that it is not clear what `obj` is. It looks like you have addressed the other part of my original concern (i.e., how `obj` relates to the other "obj" in the nearby context).

In general, `obj` means nothing. It is the same as calling something `foo` or `bar`. Everything is an object. Calling something a "hash" or "hash object" is not much better because it reflects the internal structure of the object rather than its meaning. I do not know what getCommit() returns, but I hope you can find a more descriptive name for it. If you cannot, please describe the return value, and I will try to come up with some descriptive names.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+           prParams.number = prContext.pr.number;
+           let prPromise = getPR(prParams);
+
+           return Promise.all([commitPromise, prPromise]);
```
It feels a little odd that we use .then.then.then for some promises but .all for these two. AFAICT, `prPromise` does not depend on `commitPromise` or even `sha`. Please check whether it is possible to convert this .all into .then.then.

eduard-bagdasaryan replied:
The general idea is the following: run parallel requests where applicable. You are right that prPromise and commitPromise are independent (the first gets information about the PR in question, the second - about the 'merging' commit information), and that is the exact reason why I used Promise.all(). Obviously, the following step, examining results of these two, uses 'then'.

rousskov replied:
If the rest of the related code is still (and should be) using the then.then. approach, then please convert this .all into .then.then as well.

Parallel execution is a performance optimization that often hurts code comprehension, complicates code refactoring/enhancement, increases bug risks, and even elevates GitHub query limits concerns. We are currently focusing on writing comprehensible, easy-to-enhance, correct, and robust code. Performance optimizations will come later.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+           prContext.tagSha = sha;
+           let commitPromise = getCommit(params);
+
+           let prParams = commonParams();
```
The difference between `params` and `prParams` is not clear. The former or both need better names.

eduard-bagdasaryan replied:
Fixed.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+               processNextPR();
+               return Promise.resolve("Some auto_branch checks failed for PR" + prContext.pr.number);
+           }
+           return Promise.reject("all checks succeeded");
```
Rejecting a promise on a clearly successful outcome does not make sense, even if it works. If such a rejection is needed to make the current code work, then the code should be refactored to avoid the need for this hack.

eduard-bagdasaryan replied:
The overall logic of my 'checkBeforeStartMerging()' makes the following sense to me:  try to check whether we should avoid merging for some reason. If there is a reason, or we can't calculate that reason(an exception emerged), we start merging as usual. The natural way to make this working is to use catch block for both these situations.

rousskov replied:
Please change the overall logic to a sequence of steps towards merging, with exceptions or rejections used to handle merge precondition violations and other problems. AFAICT, once you do that, then "the natural way to make this working" will itself be natural.

In other words, if the correct solution is unnatural, then the wrong problem has been solved.

rousskov replied:
>  try to check whether we should avoid merging for some reason. If [...] we can't calculate that reason(an exception emerged), we start merging as usual.

This part of the original logic seems wrong to me.

* In negative logic: If we cannot be sure that merging is not prohibited, then we should _not_ merge.
* In positive logic: If we cannot be sure that merging is allowed, then we should _not_ merge.

No matter which way you prefer to look at the problem, an exception means "no merging" rather than "merging as usual".

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+            return;
+        }
+        console.log("PR" + params.number + ": Github still calculates mergeable flag, will retry in " + delay + " msec delay");
+        setTimeout(getMergeablePR, delay, params, resolve, reject);
```
This "recursive" timeout smells like a potentially "infinite" and pretty tight loop if GitHub does not calculate the flag for some reason. I understand that a timeout and retries are needed because there may be no other event to tickle EggTimer when GitHub is done calculating the mergeable flag. However, the overall logic here should be different IMO: Wait zero seconds, a minute, and 5 minutes. If GitHub has not calculated the result after all that wait, mark PR as failed to be auto-merged, reject the promise, and move on. Using `await` may help keep the code simple. Here is a rough sketch:

```
const waits = [ 0, 1*60*1000, 5*60*1000 ]; // ms
for (wait of waits) {
    if (wait) {
        log(...);
        await promiseToWait(wait);
    }
    ... check mergeable; resolve/reject and exit if it was calculated ...
}
... throw a "GitHub cannot calculate mergeable status" error (which will label the PR)...
```

eduard-bagdasaryan replied:
```
However, the overall logic here should be different IMO: Wait zero seconds, a minute, and 5 minutes.
```

In my tests, Github calculates 'merge' result in ~1 second. Probably we should limit these polling by several seconds, I don't think we should wait 1 minute or 5 minutes(BTW, how did you calculate these times?). Also, Github documentation made me feel that this flag is calculated very quickly(though it did not say how long).

rousskov replied:
We need to balance the cost of waiting longer than necessary, the cost of retrying often, and the cost of marking the pull request as stuck/requiring human intervention. I believe my made up delays do that while repeated 500ms retries do not. I am sure there are other made up delays that would also work well -- my only source of information was your choice of a 500ms delay.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+                tagParams.ref = "refs/" + mergingTag(prContext.pr.number);
+                tagPromise = createReference(tagParams);
+            } else {
+                tagParams.ref = mergingTag(prContext.pr.number);
```
How would a mergingTag() caller know whether to prefix the result with "refs/" (like a few lines above) or not (like on this line)? Please fix the mergingTag() API.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+            if (markedAsFailed(labels))
+                console.log("PR" + prContext.pr.number + " previous merge was unsuccessful");
+
+            checkBeforeStartMerging(prContext);
```
This result-ignoring checkBeforeStartMerging() call that is not followed by some kind of merge() call is very confusing! I eventually found startMerging() inside checkBeforeStartMerging() but that is not an acceptable design. We should probably be doing some .then.then chaining here. These problems may be related to the reject-on-success weirdness inside checkBeforeStartMerging().

eduard-bagdasaryan replied:
```
This result-ignoring checkBeforeStartMerging() call that is not followed by some kind of merge() call is very confusing!
```
Since checkBeforeStartMerging() plans some async work, so, in general(promises aside), I don't see what kind of confusion it can make. Probably it could be named better, e.g., something like 'checkAdditionalConditionsAndStartIfOk()", to better describe what it really does, but this looks too long for the method name. Probably we could promisify checkBeforeStartMerging() itself (as well as other few high-level methods), creating additional promise-level, simplifying in one place but complicating in other.

rousskov replied:
I believe I have already documented what confused me.

Yes, the function is misnamed, but I doubt that adding AndStart suffix to the name is the best or even correct solution here because I suspect that this necessity to add AndFooAndBarAndBaz suffixes is an overall design problem, not a naming problem. With the right design, this naming problem would not exist.

rousskov commented on eggtimer.js:
```diff
@@ -1,326 +1,681 @@
+   refParams.ref = mergingTag(prContext.pr.number);
+   let tagTreeSha = null;
+   prContext.tagSha = null;
+   console.log("Checking whether to skip due to a previous unsuccessful merge for PR" + prContext.pr.number);
```
The code below appears to be doing more than that. For example it also checks test statuses. In fact, it is difficult for me to find the place where we "skip due to a previous unsuccessful merge".

eduard-bagdasaryan replied:
I [made](https://github.com/measurement-factory/EggTimer/pull/1#discussion_r153134008) a quick explanation of 'checkBeforeStartMerging()' purpose. Yes, it checks the 'merging' commit statuses(which could have been changed) to make the appropriate decision.

rousskov replied:
Please note that explaining checkBeforeStartMerging() purpose does nothing to address my concern, even if I completely agree with the explained purpose: When I or somebody else looks at this code a year from now, they will have the same concern.

rousskov reviewed (changes requested):
This review is based on commit https://github.com/measurement-factory/EggTimer/pull/1/commits/73d90c73aa26d1cfcc61dd4d1c4f9aacfb61f9e9. Two more commits were added since I started looking at the code. I did not look at them (yet).

I got overwhelmed with code/design problems and could not finish the review. I hope the inlined review comments will help us move forward, but there are probably more problems waiting to be discovered here.

Is there a better way to _test_ EggTimer than deploying it on some sandbox repository, manually creating a few PRs, and trying to understand whether the bot processed those PRs correctly?

eduard-bagdasaryan commented:
```
Is there a better way to test EggTimer than deploying it on some sandbox repository, manually creating a few PRs, and trying to understand whether the bot processed those PRs correctly?
```

Some testing could be made via my 'dry_run' option on. With it, the bot does not do modifying operations, it just receives events and analyzes merge preconditions.
A better way is to construct the required testing environment ourselves, i.e. create automatically a test Github repository from scratch and perform some pre-defined operations (such as creating and approving PRs) and check whether the bot handles all this correctly. I doubt this could be done quickly easily.

rousskov commented:
FYI: When responding to GitHub comments, if you quote text using `>` instead of triple backquotes, then the quoted text will wrap normally. A single `>` is also easier to type :\-).

rousskov commented:
BTW, feel free to close this pull request and (when ready) create a new one. Due to significant code changes, I doubt the old comments will remain relevant anyway. And we know they will be difficult to find. Your call.

eduard-bagdasaryan committed:
Addressed several review comments.

eduard-bagdasaryan committed:
Big refactoring, according to the review and discussion.

The major work completed, few tests passed.

There are still few not_addressed/not_completed questions,
such as configuration improvement and better error handling.

eduard-bagdasaryan committed:
Big refactoring, according to the review and discussion, pt. 2.

This change addresses questions, not completed in f5d9f72, such
as creating class wrappers for 'Config' or 'Context' entities,
simplifying logging, and fixing some found problems.

eduard-bagdasaryan committed:
Properly handle possible getPRList() rejections.

eduard-bagdasaryan committed:
Code improvements after review:

1. Get PR list in runStep(), instead of run(). runStep() algorithm is
   the following now: select PRs from the list one by one and process
   them, until either:

   * there is a PR we need to wait to,
   * we processed all PRs from the list,
   * an unexpected error occurred.

2. Protect ConfigOptions fields from typos: use 'public' methods
   instead.

3. Re-run in 10 minutes (default) in a case of unexpected error.

eduard-bagdasaryan committed:
Several improvements after discussion.

* Get 'required' status checks from Github (instead of configuring
  required status checks number).

* Get reviewers list from Github. For now, the bot keeps using the
  existing 'reviewers_number' parameter, but analyses the reviewers
  having 'push' right (these are considered as valid approvals).

* Use 'rerun' flag in 'run' instead of 'runStep' for simplification.

* A few other fixes, e.g., missed 'await' in cleanupOnError().

eduard-bagdasaryan committed:
Code refactoring.

* Made exception handling style more consistent:
  handle most of exceptions in high-level runStep(), instead of
  startMerging(), finishMerging(), etc. Made cleanupOnError()
  non-throw for this purpose.

* Distinguish FF-merge failure, checking for 422 HTTP error. Re-run
  merge, if such PR was the last one in the list.

* getPR() optimization: do not wait while Github calculates
  'mergeable' flag, if we do not need it(yet).

* Polishing: removed some stale code, renamed RejectMsg -> ErrorContext,
  etc.

rousskov commented:
Closing as stale, eventually replaced by https://github.com/measurement-factory/anubis/pull/25.
